test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package benchmark
+
+import (
+	"fmt"
+	"io/ioutil"
+	v1 "k8s.io/api/core/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/component-base/featuregate"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/test/integration/framework"
+	testutils "k8s.io/kubernetes/test/utils"
+	"sigs.k8s.io/yaml"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+var (
+	configFile = "config/performance-config.yaml"
+)
+
+// TestCase configures a test case to run the scheduler performance test. Users should be able to
+// provide this via a YAML file.
+//
+// It specifies nodes and pods in the cluster before running the test. It also specifies the pods to
+// schedule during the test. The config can be as simple as just specify number of nodes/pods, where
+// default spec will be applied. It also allows the user to specify a pod spec template for more compicated
+// test cases.
+//
+// It also specifies the metrics to be collected after the test. If nothing is specified, default metrics
+// such as scheduling throughput and latencies will be collected.
+type TestCase struct {
+	// description of the test case
+	Desc string
+	// configures nodes in the cluster
+	Nodes nodeCase
+	// configures pods in the cluster before running the tests
+	InitPods podCase
+	// pods to be scheduled during the test.
+	PodsToSchedule podCase
+	// optional, feature gates to set before running the test
+	FeatureGates map[featuregate.Feature]bool
+}
+
+type nodeCase struct {
+	Num                      int
+	NodeTemplatePath         *string
+	NodeAllocatableStrategy  *testutils.NodeAllocatableStrategy
+	LabelNodePrepareStrategy *testutils.LabelNodePrepareStrategy
+}
+
+type podCase struct {
+	Num                               int
+	PodTemplatePath                   *string
+	ReplicationControllerName         *string
+	PersistentVolumeTemplatePath      *string
+	PersistentVolumeClaimTemplatePath *string
+}
+
+func TestSchedulerPerformance(t *testing.T) {
+	var tests []TestCase
+	getSpec(&configFile, &tests)
+
+	for _, test := range tests {
+		name := fmt.Sprintf("%v/%vNodes/%vPods", test.Desc, test.Nodes.Num, test.InitPods.Num)
+		t.Run(name, func(t *testing.T) {
+			for feature, flag := range test.FeatureGates {
+				defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, feature, flag)()
```

Comment:
Why defer? Shouldn't the feature gates be set before running the tests?

Reply:
Please note there is `()` at the end of this line, we are calling the featuregatetesting.SetFeatureGateDuringTest to get the defer function, the feature gates are actually set here,  and the defer function is called to reset the feature flags at the end of each tests.

Reply:
Ack. Maybe more clear if you do 
```
resetFunc := featuregatetesting.SetFeatureGateDuringTest
defer resetFunc()
```

But this is also fine.